axelor-supplychain/src/main/java/com/axelor/apps/supplychain/web/SaleOrderController.java

```diff
@@ -304,6 +286,97 @@ public void generateInvoice(ActionRequest request, ActionResponse response) {
+          qtyToInvoiceMap.put(soLineId, qtyToInvoiceItem);
+          BigDecimal priceItem = new BigDecimal(map.get(SO_LINES_WIZARD_PRICE_FIELD).toString());
+          priceMap.put(soLineId, priceItem);
+          BigDecimal qtyItem = new BigDecimal(map.get(SO_LINES_WIZARD_QTY_FIELD).toString());
+          qtyMap.put(soLineId, qtyItem);
+        }
+      }
+    }
+  }
+
+  public void generateInvoicesFromSelectedLines(ActionRequest request, ActionResponse response) {
+
+    try {
+
+      SaleOrderRepository saleOrderRepository = Beans.get(SaleOrderRepository.class);
+      Context context = request.getContext();
+      List<Map<String, Object>> saleOrderLineListContext;
+      saleOrderLineListContext =
+          (List<Map<String, Object>>) request.getRawContext().get("saleOrderLineListToInvoice");
+      if (saleOrderLineListContext.isEmpty()) {
+        response.setAlert(I18n.get("No items have been selected."));
+        return;
+      }
+      int operationSelect = SaleOrderRepository.INVOICE_LINES;
+      boolean isPercent = (Boolean) context.getOrDefault("isPercent", false);
+
+      Map<SaleOrder, List<Map<String, Object>>> saleOrderLineListContextMap =
+          saleOrderLineListContext.stream()
+              .collect(
+                  groupingBy(
+                      stringObjectMap ->
+                          saleOrderRepository.find(
+                              Long.valueOf(
+                                  (Integer)
+                                      ((LinkedHashMap<?, ?>) stringObjectMap.get("saleOrder"))
+                                          .get("id")))));
+
+      Map<SaleOrder, BigDecimal> amountToInvoiceMap = new HashMap<>();
+      Map<SaleOrder, Map<Long, BigDecimal>> qtyMaps = new HashMap<>();
+      Map<SaleOrder, Map<Long, BigDecimal>> qtyToInvoiceMaps = new HashMap<>();
+      Map<SaleOrder, Map<Long, BigDecimal>> priceMaps = new HashMap<>();
+      for (Map.Entry<SaleOrder, List<Map<String, Object>>> entry :
+          saleOrderLineListContextMap.entrySet()) {
+        SaleOrder saleOrder = entry.getKey();
+        BigDecimal amountToInvoice =
+            saleOrder.getExTaxTotal().subtract(saleOrder.getAmountInvoiced());
+        amountToInvoiceMap.put(saleOrder, amountToInvoice);
+        Map<Long, BigDecimal> qtyMap = new HashMap<>();
+        Map<Long, BigDecimal> qtyToInvoiceMap = new HashMap<>();
+        Map<Long, BigDecimal> priceMap = new HashMap<>();
+        fillMaps(entry.getValue(), qtyToInvoiceMap, priceMap, qtyMap);
+        qtyMaps.put(saleOrder, qtyMap);
+        qtyToInvoiceMaps.put(saleOrder, qtyToInvoiceMap);
+        priceMaps.put(saleOrder, priceMap);
+      }
+
+      Beans.get(SaleOrderInvoiceService.class)
+          .generateInvoicesFromSOL(
+              priceMaps, qtyToInvoiceMaps, qtyMaps, amountToInvoiceMap, isPercent, operationSelect);
+      response.setCanClose(true);
+
+    } catch (Exception e) {
+      TraceBackService.trace(response, e);
+    }
```

Comment:
If the exception is an axelorException, you can just throw it here in the controller (it will display an error message in the popup)